videostore/indexer.go

```diff
@@ -0,0 +1,451 @@
+	}
+	ix.setupDone = true
+	return nil
+}
+
+func (ix *indexer) initializeDB() error {
+	query := fmt.Sprintf(`
+	CREATE TABLE IF NOT EXISTS %s (
+		file_path TEXT PRIMARY KEY,
+		start_time_unix INTEGER NOT NULL,
+		duration_ms INTEGER NOT NULL,
+		size_bytes INTEGER NOT NULL,
+		inserted_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP
+	);
+	CREATE INDEX IF NOT EXISTS idx_start_time ON %s (start_time_unix);
+	`, segmentsTableName, segmentsTableName)
+
+	_, err := ix.db.Exec(query)
+	return err
+}
+
+func (ix *indexer) run(ctx context.Context) {
+	ix.logger.Debug("starting Indexer polling loop")
+	ticker := time.NewTicker(indexerPollingInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			ix.logger.Debug("indexer polling loop has stopped")
+			return
+		case <-ticker.C:
+			if err := ix.cleanupAndIndex(ctx); err != nil {
+				ix.logger.Errorw("error during indexer maintenance", "error", err)
+			}
+		}
+	}
+}
+
+func (ix *indexer) cleanupAndIndex(ctx context.Context) error {
+	// 1. Clean up db according to storage limits
+	deletedFilePaths, err := ix.cleanupDB(ctx)
+	if err != nil {
+		ix.logger.Errorw("failed to cleanup db", "error", err)
+	}
+
+	// 2. Delete disk files that were just deleted from db
+	for _, path := range deletedFilePaths {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if err := os.Remove(path); err == nil {
+			ix.logger.Infof("deleted file from disk: %s", path)
+		} else {
+			ix.logger.Warnw("failed to delete file from disk", "file", path, "error", err)
+		}
+	}
+
+	// 3. Index new files
+	if err := ix.indexNewFiles(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// indexNewFiles indexes new video files on disk.
+func (ix *indexer) indexNewFiles(ctx context.Context) error {
+	diskFiles, err := ix.getDiskFiles(ctx)
+	if err != nil {
+		return err
+	}
+	indexedFiles, err := ix.getIndexedFiles(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Query for the oldest DB file in order to determine if a file is new
+	var oldestDBFile string
+	row := ix.db.QueryRowContext(ctx, fmt.Sprintf("SELECT file_path FROM %s ORDER BY start_time_unix ASC LIMIT 1", segmentsTableName))
+	if err := row.Scan(&oldestDBFile); err != nil && err != sql.ErrNoRows {
+		return fmt.Errorf("failed to get oldest DB entry: %w", err)
+	}
+	var oldestDBTime time.Time
+	if oldestDBFile != "" {
+		oldestDBTime, err = extractDateTimeFromFilename(oldestDBFile)
+		if err != nil {
+			return fmt.Errorf("failed to parse filename of oldest db entry %q: %w", oldestDBFile, err)
+		}
+	}
+
+	for path, info := range diskFiles {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+		if _, exists := indexedFiles[path]; exists {
+			continue
+		}
+		if !oldestDBTime.IsZero() {
+			fileTime, err := extractDateTimeFromFilename(path)
+			if err != nil {
+				return fmt.Errorf("failed to extract timestamp from filename, skipping: %w", err)
+			}
+			if fileTime.Before(oldestDBTime) {
```

Comment:
so... consider the following case:
1. the module starts indexing
2. it get half way through the files on disk
3. it crashes
4. it reboots
5. It never indexes the files that were on disk when it booted which it didn't index yet...

Reply:
Synced offline and arrived at the current solution of marking as deleted in the db then actually deleting later